src/compiler/emitter.ts

```diff
@@ -144,81 +147,193 @@ module ts {
+                    case SyntaxKind.ExportDeclaration:
+                        return generateNameForExportDeclaration(<ExportDeclaration>node);
+                    case SyntaxKind.ExportAssignment:
+                        return generateNameForExportAssignment(<ExportAssignment>node);
+                    default: 
+                        Debug.fail("Unexpected node kind:" + node.kind);
+                        return undefined;
                 }
             }
 
-            function generateUniqueNameForLocation(location: Node, baseName: string): string {
-                let name: string
-                // first try to check if base name can be used as is
-                if (!isExistingName(location, baseName)) {
-                    name = baseName;
+            function isUniqueName(name: string) {
+                return !resolver.hasGlobalName(name) &&
+                    !hasProperty(currentSourceFile.identifiers, name) &&
+                    (!generatedNames || !hasProperty(generatedNames, name));
+            }
+
+            function isUniqueTempVariableName(location: Node, name: string): boolean {
+                if (!resolver.isUnknownIdentifier(location, name)) {
+                    return false;
                 }
-                else {
-                    name = generateUniqueName(baseName, n => isExistingName(location, n));
+
+                return !currentScopeTemp.names || !hasProperty(currentScopeTemp.names, name);
+            }
+
+            function makeUniqueName(baseName: string): string {
+                // First try '_name'
+                if (baseName.charCodeAt(0) !== CharacterCodes._) {
+                    baseName = "_" + baseName;
+                    if (isUniqueName(baseName)) {
+                        return recordGeneratedName(baseName, /*isTempVariableName*/ false);
+                    }
+                }
+                // Find the first unique '_name_n', where n is a positive number
+                if (baseName.charCodeAt(baseName.length - 1) !== CharacterCodes._) {
+                    baseName += "_";
+                }
+                let i = 1;
+                while (true) {
+                    let name = baseName + i;
+                    if (isUniqueName(name)) {
+                        return recordGeneratedName(name, /*isTempVariableName*/ false);;
+                    }
+                    i++;
                 }
-                
-                return recordNameInCurrentScope(name);
             }
-            
-            function recordNameInCurrentScope(name: string): string {
-                if (!currentScopeNames) {
-                    currentScopeNames = {};
+
+            // Create a temporary variable with a unique unused name.
+            function createTempVariable(location: Node, preferredName?: string): Identifier {
+                for (var name = preferredName; !name || !isUniqueTempVariableName(location, name); currentScopeTemp.counter++) {

```

Comment:
Doesn't it have to be?  'preferredName' is optional.  So 'name' may be undefined.  And def want to run the loop in that case.